docs/adr/0005-arc-pointers.md

```diff
@@ -0,0 +1,167 @@
+# Use raw `Arc<T>` pointers to pass objects across the FFI
+
+* Status: proposed
+* Deciders: mhammond, rfkelly
+* Consulted: travis, jhugman, dmose
+* Date: 2021-04-19
+
+Discussion and approval: [PR 430](https://github.com/mozilla/uniffi-rs/pull/430).
+
+Technical Story: [Issue 419](https://github.com/mozilla/uniffi-rs/issues/419).
+
+Prototype: [PR 420](https://github.com/mozilla/uniffi-rs/pull/420).
+
+## Context and Problem Statement
+
+UniFFI currently manages object instances using the `HandleMap` struct in the ffi-support crate.
+This means that external consumers of UniFFI-wrapped interfaces never see
+any pointers to structs - instead, they get what is (roughly) an index into
+an array, with the struct being stored in (and owned by) that array.
+
+This has a number of safety characteristics which are particularly important for
+hand-written FFI interfaces, but it does cause some issues in evolving UniFFI in
+directions we consider important. In addition to the slight performance overhead,
+the use of `HandleMap`s makes it difficult to support:
+
+* Passing object instances as arguments ([#40](https://github.com/mozilla/uniffi-rs/issues/40)).
+  Getting objects out of a `HandleMap` involves a closure, so accepting multiple
+  object-typed arguments would involve code-generating nesture closures.
+* Returning object instances from functions ([#197](https://github.com/mozilla/uniffi-rs/issues/197)).
+  Does the returned object already exist in the handlemap? If so, what is its handle?
+  How will we manage the lifetime of multiple references to the object?
```

Comment:
We could add another bullet point here about having object-typed fields in records, which AFAICT is going to Just Work as part of this approach.